### PR TITLE
Use parameters for info addon

### PR DESCRIPTION
**Pull request by Keraito**
Last of #3625 

On top of implementing the new parameterized way of using addons for `info`, I also added made changes regarding how the options and parameters were processed with respect to each other, https://github.com/storybooks/storybook/issues/3625#issuecomment-393107013. Before, the addon parameters would be the ones that were the latest set (global -> local -> story), replacing the previous ones entirely. 

Now, it only replaces the addon properties per key, https://github.com/storybooks/storybook/issues/3625#issuecomment-393600705. I don't think my solution in `lib/core/src/client/preview/client_api.js` is the cleanest one, so if anyone else has a better idea, happy to hear! I tried some stuff with `reduce` but that only made more of a mess. 

I applied similar logic to the handling of `options` and `parameters`.

## Core tests
`2 snapshot tests failed in 1 test suite`, both output: 
```
    - Snapshot
    + Received
    
...

    - <Component>
    + <deprecated>
```

This seems unrelated to my changes, or am I mistaken?

### CODE REVIEW

File: addons/info/README.md
Comment (tmeasday) [inline addons/info/README.md]: I think it is worth indicating this should be done in `.storybook/config.js`

File: addons/info/README.md
Comment (tmeasday) [inline addons/info/README.md]: Let's deprecate this and the usage as a HoC ("Basic usage" and "Usage with options") -- so let's remove it from the docs.

File: addons/info/src/index.js
Comment (tmeasday) [inline addons/info/src/index.js]: Do we want to skip the addon if they don't set options or parameters? Either by passing `skipIfNoOptionsOrParameters: true` or manually doing it here..

**Review by tmeasday — COMMENTED** [review]
Some comments to remove the old API a little more.

I'm not sure if `info` is a special case because it needs to be the first addon?

File: lib/core/src/client/preview/client_api.js
Comment (tmeasday) [inline lib/core/src/client/preview/client_api.js]: Maybe just `{ ...allParam[key], ...params[key] }`? Kind of odd to use `Object.assign` and object spread ;)

**Comment by tmeasday** [comment]
Look great btw! Thanks @Keraito

File: addons/info/README.md
Comment (shilman) [reply addons/info/README.md]: @tmeasday on a related note, do you think we should have a `Deprecated` section in `storybook-official` so we still get story snapshots for deprecated usage, but they are separate from the recommended usage examples?

File: addons/info/src/index.js
Comment (Keraito) [reply addons/info/src/index.js]: Def agree with the `disable: true` part, will do. I'm not so sure about the `skipIfNoOptionsOrParameters` part though. I assume the both of them do the same thing, namely disable the addon entirely. But `info` is slightly different from the others addons imo that it does something even without configuration: display Proptypes, the component, etc.

Example: https://github.com/storybooks/storybook/pull/3697/files#diff-c915ecbefba3a2734ebd93eb258431dbR353 This is basically an parameters- and options-less story, besides some minor options (header styling) and parameters (inline and manual text). The configuration don't add a lot, but the addon itself adds a lot of information stand-alone already.

So I would say just only documenting the disable flag should be the only thing

File: addons/info/README.md
Comment (Keraito) [reply addons/info/README.md]: @tmeasday Just so we're on the same line: you want to deprecate the `setDefaults` way of setting defaults (and shift that to the `addDecorator` in the `config.js` with either options or parameters) and remove all the docs involving this + the HoC?

Related to this, should I then rewrite all the existing `info` stories in `official-storybook` to use the parameterized way?

**Comment by Keraito** [comment]
I processed most of the feedback. Waiting on a reply on https://github.com/storybooks/storybook/pull/3697#discussion_r192557635, then i will finish everything up

File: addons/info/src/index.js
Comment (tmeasday) [reply addons/info/src/index.js]: Ok, makes sense. Thanks for thinking about that.

**Comment by Keraito** [comment]
Oops I totally referenced the wrong comment in my last message, I meant this one https://github.com/storybooks/storybook/pull/3697#discussion_r192557792 (about deprecating the previous usage and what to change exactly). 🙈sorry @tmeasday could you take a look again?

File: addons/info/README.md
Comment (tmeasday) [reply addons/info/README.md]: Sorry guys, missed this!

> should have a Deprecated section in storybook-official so we still get story snapshots for deprecated usage, but they are separate from the recommended usage examples?

Probably a good idea @shilman -- perhaps a small refactor of the stories after we get the last few PRs merged is in order.

> Just so we're on the same line: you want to deprecate the setDefaults way of setting defaults (and shift that to the addDecorator in the config.js with either options or parameters) and remove all the docs involving this + the HoC?

Yeah that was what I was thinking, do you agree? Definitely get rid of the HoC; and `setDefaults` seems redundant now, right?

> Related to this, should I then rewrite all the existing info stories in official-storybook to use the parameterized way?

What I've been doing is leaving the existing ones as "tests" for the deprecated functionality (we can remove them when actually delete that functionality. Then making sure there are stories to cover all the new ways of doing it! You don't need to re-test that parameter inheritance works though (although maybe adding a the story in `core.stories.js` to test your change to the way it works?), just add stories that make sense.

I hope I'm making sense here 😬

File: addons/info/src/index.js
Comment (Keraito) [inline addons/info/src/index.js]: I don't feel like this is how I should've implemented the depreciation, but only this worked for me for some reason.

**Comment by Keraito** [comment]
Should be all done now, could u review again @tmeasday ! 😊

**Review by tmeasday — APPROVED** [review]
Looks great @Keraito!

**Comment by Hypnosphi** [comment]
Please update test snapshots: `yarn test --core --update`

**Comment by Keraito** [comment]
I'm kinda confused at what's going on. I actually updated the storyshots before (https://github.com/storybooks/storybook/pull/3697/commits/29aaa627cc00a32adc1916d994dcfd30d56488fe). My second commit (after running `yarn`) removed all the updates to the snapshots. The first run of `yarn test --core --update` however also removed the snapshots update of my added stories and the error described in https://github.com/storybooks/storybook/pull/3697#issuecomment-395704000 is still present

**Comment by Keraito** [comment]
Nevermind all of that, apperently I just didn't have the dev server running 🤦‍♂️ Should be all good now

**Pull request merged**
